pkg/utils/cluster.go

```diff
@@ -161,3 +171,94 @@ func AnnotateNode(ctx context.Context, nodeName string, key, value string, c cli
 
 	return AnnotateObject(ctx, node, key, value, c)
 }
+
+func FindNodePoolConfig(ctx context.Context, node *corev1.Node, c client.Client) (*sriovnetworkv1.SriovNetworkPoolConfig, []corev1.Node, error) {
+	logger := log.FromContext(ctx)
+	logger.Info("FindNodePoolConfig():")
+	// get all the sriov network pool configs
+	npcl := &sriovnetworkv1.SriovNetworkPoolConfigList{}
+	err := c.List(ctx, npcl)
+	if err != nil {
+		logger.Error(err, "failed to list sriovNetworkPoolConfig")
+		return nil, nil, err
+	}
+
+	selectedNpcl := []*sriovnetworkv1.SriovNetworkPoolConfig{}
+	nodesInPools := map[string]interface{}{}
+
+	for _, npc := range npcl.Items {
+		// we skip hw offload objects
+		if npc.Spec.OvsHardwareOffloadConfig.Name != "" {
+			continue
+		}
+
+		if npc.Spec.NodeSelector == nil {
+			npc.Spec.NodeSelector = &metav1.LabelSelector{}
+		}
+
+		selector, err := metav1.LabelSelectorAsSelector(npc.Spec.NodeSelector)
+		if err != nil {
+			logger.Error(err, "failed to create label selector from nodeSelector", "nodeSelector", npc.Spec.NodeSelector)
+			return nil, nil, err
+		}
+
+		if selector.Matches(labels.Set(node.Labels)) {
+			selectedNpcl = append(selectedNpcl, npc.DeepCopy())
+		}
+
+		nodeList := &corev1.NodeList{}
+		err = c.List(ctx, nodeList, &client.ListOptions{LabelSelector: selector})
+		if err != nil {
+			logger.Error(err, "failed to list all the nodes matching the pool with label selector from nodeSelector",
+				"machineConfigPoolName", npc,
+				"nodeSelector", npc.Spec.NodeSelector)
+			return nil, nil, err
+		}
+
+		for _, nodeName := range nodeList.Items {
+			nodesInPools[nodeName.Name] = nil
+		}
+	}
+
+	if len(selectedNpcl) > 1 {
+		// don't allow the node to be part of multiple pools
+		err = fmt.Errorf("node is part of more then one pool")
+		logger.Error(err, "multiple pools founded for a specific node", "numberOfPools", len(selectedNpcl), "pools", selectedNpcl)
+		return nil, nil, err
+	} else if len(selectedNpcl) == 1 {
+		// found one pool for our node
+		logger.V(2).Info("found sriovNetworkPool", "pool", *selectedNpcl[0])
+		selector, err := metav1.LabelSelectorAsSelector(selectedNpcl[0].Spec.NodeSelector)
+		if err != nil {
+			logger.Error(err, "failed to create label selector from nodeSelector", "nodeSelector", selectedNpcl[0].Spec.NodeSelector)
+			return nil, nil, err
+		}
+
+		// list all the nodes that are also part of this pool and return them
```

Comment:
for those nodes why arent we validating they match exactly one ncp ? like in L223